### PR TITLE
Massive refactoring

### PR DESCRIPTION
**Pull request by descorp**
### Changes

* rename FullFormAddressItem -> FormAddressItem
* rename AddressInfo -> PostalAddress
* rename Switch -> Toggle
* make `Amount` a standalone object
* make countryCode and currencyCode non-optional and assert validation on init
* new file for ShopperInformation
* new file for DisplayInformation

### Folder structure

Screenshot 2021-05-21 at 18 44 23

### CODE REVIEW

File: Adyen/Model/Payment.swift [hunk 1]
```diff
@@ -0,0 +1,28 @@
+//
+// Copyright (c) 2021 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+/// Describes the current payment.
+public struct Payment {
```
Comment (kaphacius) [inline Adyen/Model/Payment.swift:10]: can `amount` and `country` code be constants?

File: AdyenActions/Components/Base/ActionComponent.swift [hunk 2]
```diff
@@ -15,6 +15,13 @@ public protocol ActionComponent: Component {
     
 }
 
+/// Represents any structure/class that can be initialized without any parameters.
+/// :nodoc:
+public protocol ParameterlessInitializable {
```
[the remark this answers is not in the document]
Comment (descorp) [reply AdyenActions/Components/Base/ActionComponent.swift:20]: some magic for WeChat's "decoupled" init

File: Adyen/Core/Payment Methods/DisplayInformation.swift [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        self.title = title
+        self.subtitle = subtitle
+        self.logoName = logoName
+        self.disclosureText = disclosureText
+        self.footnoteText = footnoteText
+    }
+}
+
+/// :nodoc:
+extension DisplayInformation: Equatable {}
```
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/Core/Payment Methods/DisplayInformation.swift:55]: Just copied this code. Why not 🤔

File: Adyen/Formatters/IBANFormatter.swift [hunk 4]
```diff
@@ -10,7 +10,7 @@ import Foundation
 public final class IBANFormatter: Formatter {
     
     /// Initializes the IBAN formatter.
-    public init() {}
+    public init() { /* public init */ }
```
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/Formatters/IBANFormatter.swift]: Removed this rule from Sonar!

File: Adyen/Core/Components/AbstractPersonalInformationComponent/AbstractPersonalInformationComponent.swift [hunk 5]
```diff
@@ -141,8 +141,7 @@ open class AbstractPersonalInformationComponent: PaymentComponent, PresentableCo
     public var phoneItem: FormPhoneNumberItem? { phoneItemInjector?.item }
 
     private lazy var selectableValues: [PhoneExtensionPickerItem] = {
-        let query = PhoneExtensionsQuery(paymentMethod: PhoneNumberPaymentMethod.mbWay)
-        return getPhoneExtensions().map {
+        getPhoneExtensions().map {
```
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/Core/Components/AbstractPersonalInformationComponent/AbstractPersonalInformationComponent.swift:159]: I have a better idea :D
`PhoneExtensionViewModel` is deprecated now

File: Adyen/Helpers/StringHelpers.swift [hunk 6]
```diff
@@ -33,6 +33,15 @@ public extension AdyenScope where Base == String? {
 /// :nodoc:
 public extension AdyenScope where Base == String {
 
+    /// Return flag emoji if string is a country code; otherwise returns emty string.
+    var countryFlag: String {
```
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/Helpers/StringHelpers.swift:37]: 👍

File: Adyen/Model/Amount.swift [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+/// :nodoc:
+public extension Amount {
+
+    /// Returns a formatter representation of the amount.
+    ///
+    /// :nodoc:
+    var formatted: String {
+        guard let formattedAmount = AmountFormatter.formatted(amount: value, currencyCode: currencyCode) else {
```
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/Model/Amount.swift:51]: 👍

File: Adyen/Model/Payment.swift (hunk 1 above)
Comment (descorp) [reply Adyen/Model/Payment.swift:10]: 👍

File: Adyen/Validators/CountryCodeValidator.swift [hunk 8]
```diff
@@ -7,19 +7,21 @@
...
 public struct CountryCodeValidator: Validator {
 
+    private static let allCountryCodes = Set(Locale.isoRegionCodes)
+
     /// :nodoc:
-    public init() { /* Empty initializer */ }
+    public init() {}
 
     /// :nodoc:
     public func isValid(_ value: String) -> Bool {
```
[the remark this answers is not in the document]
Comment (descorp) [reply Adyen/Validators/CountryCodeValidator.swift:19]: It is part of `Validator ` protocol. Maybe in other PR :D

File: AdyenActions/Components/Base/ActionComponent.swift (hunk 2 above)
Comment (mohammedDehairy) [inline AdyenActions/Components/Base/ActionComponent.swift:20]: can we make this internal?

File: Adyen/Model/BrowserInfo.swift [hunk 9]
```diff
@@ -25,7 +25,9 @@ public struct BrowserInfo: Encodable {
         webView = WKWebView()
         webView?.evaluateJavaScript("navigator.userAgent") { result, _ in
             webView = nil
-            guard let result = result as? String else { completion(nil); return }
+            guard let result = result as? String else {
+                return completion(nil)
+            }
```
Comment (mohammedDehairy) [inline Adyen/Model/BrowserInfo.swift:30]: why?, one line is not long enough to make multiline

File: AdyenActions/Components/Base/ActionComponent.swift (hunk 2 above)
Comment (descorp) [reply AdyenActions/Components/Base/ActionComponent.swift:20]: No, it is used in WeChat module :(

File: Adyen/Model/BrowserInfo.swift (hunk 9 above)
Comment (descorp) [reply Adyen/Model/BrowserInfo.swift:30]: my main focuse was `completion(nil); return` ..
I can do 

``` guard let result = result as? String else { return completion(nil) } ```

File: Adyen/Core/Components/Base/PresentableComponent.swift [hunk 10]
```diff
@@ -43,7 +43,15 @@ public extension PresentableComponent {
 }
 
 /// :nodoc:
-public protocol TrackableComponent: Component, PaymentMethodAware, ViewControllerDelegate {}
+public protocol TrackableComponent: Component, PaymentMethodAware, ViewControllerDelegate {
+
+    /// :nodoc:
+    func viewDidLoad(viewController: UIViewController)
```
Comment (kaphacius) [inline Adyen/Core/Components/Base/PresentableComponent.swift:49]: did you run into the issue where you could not override default implementation of the `ViewControllerDelegate` protocol? :)

File: Adyen/Helpers/StringHelpers.swift [hunk 11]
```diff
@@ -33,6 +33,15 @@ public extension AdyenScope where Base == String? {
 /// :nodoc:
 public extension AdyenScope where Base == String {
 
+    /// Return flag emoji if string is a country code; otherwise returns emty string.
+    var countryFlag: String? {
+        guard CountryCodeValidator().isValid(base) else { return nil }
+        let baseIndex = 127397
+        var usv = String.UnicodeScalarView()
+        base.utf16.compactMap { UnicodeScalar(baseIndex + Int($0)) }.forEach { usv.append($0) }
```
Comment (kaphacius) [inline Adyen/Helpers/StringHelpers.swift:41]: maybe better to use reduce here
```swift
        return String(
            base.utf16
                .compactMap { UnicodeScalar(baseIndex + (Int ($0))) }
                .reduce(into: String.UnicodeScalarView()) { usv, scalar in
                    usv.append(scalar)
                }
            )
```

File: Adyen/Model/Amount.swift [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+/// Describes the amount of a payment.
+public struct Amount: Codable, Equatable {
+
+    /// The value of the amount in minor units.
+    public var value: Int
```
Comment (kaphacius) [inline Adyen/Model/Amount.swift:13]: can we make `value` and `currencyCode` `let`s?

Comment (descorp) [reply Adyen/Model/Amount.swift:13]: 👍

File: Adyen/Core/Components/Base/PresentableComponent.swift (hunk 10 above)
Comment (descorp) [reply Adyen/Core/Components/Base/PresentableComponent.swift]: WIP

File: Adyen/Helpers/StringHelpers.swift (hunk 11 above)
Comment (descorp) [reply Adyen/Helpers/StringHelpers.swift:41]: This is indeed nice place for reduce 👍

**Review by kaphacius — APPROVED** [review]

File: AdyenComponents/Apple Pay/ApplePayComponent.swift [hunk 13]
```diff
@@ -34,21 +34,13 @@ public class ApplePayComponent: NSObject, PresentableComponent, PaymentComponent
...
         }
         guard PKPaymentAuthorizationViewController.canMakePayments(usingNetworks: configuration.supportedNetworks) else {
             throw Error.userCannotMakePayment
         }
-        guard let countryCode = configuration.payment.countryCode, CountryCodeValidator().isValid(countryCode) else {
-            throw Error.invalidCountryCode
-        }
-        guard CurrencyCodeValidator().isValid(configuration.payment.amount.currencyCode) else {
-            throw Error.invalidCurrencyCode
-        }
```
[the remark this answers is not in the document]
Comment (descorp) [reply AdyenComponents/Apple Pay/ApplePayComponent.swift:51]: Good one!

**Review by kaphacius — APPROVED** [review]

**Review by mohammedDehairy — APPROVED** [review]

**Pull request merged**